Added more Steam filters.

BanCrash opened the pull request:
Now, Steam: Installed, Uninstalled, All, Played and False.

Please note that now Steam:true is not supported anymore because I think is duplicated with steam:installed, but If you don't think that way I can add it too.

EDITED: Added Random filter. Will show one random game every time it is executed with the conditions selected. Random:all (one random game in games installed or not), Random:steam (one random game in Steam Games installed or not) and Random:Played (one random game in All games, hidden, installed or not that have more that 0 hours of execution).

EDITED2: I am in process of adding a new feature to show hidden games according to if HiddenFilters is Enabled or Disabled in Variables.inc. If it is disabled hidden games only be show in games:all and hidden:() filters. If it is enabled, in almost everything thoug. The idea is to add an option in settings to enable or disable this.

Kapiainen commented on Lauhdutin/@Resources/Frontend/GUI.lua:
```diff
@@ -264,8 +264,58 @@ end
+			asPattern = asPattern:sub(7)
+			if StartsWith(asPattern, 'i') then
+			for i, game in ipairs(T_ALL_GAMES) do
+				if game["platform"] == 0 then
```
Fields of the `GAME_KEYS` table should be used as the key instead. If the value of that field needs to be changed in the future, then every instance of `"platform"` would have to be manually updated as it currently is. Using `GAME_KEYS.PLATFORM` as the key means that only one line of code needs to be updated, if the value needs to be changed in the future.

The same applies to `== 0`, which should be replaced with `== PLATFORM.STEAM`.

BanCrash replied:
This issue is solved too.

Kapiainen commented on Lauhdutin/@Resources/Frontend/GUI.lua:
```diff
@@ -337,6 +387,79 @@ end
 			else
 				return tResult, true
 			end	
+		elseif StartsWith(asPattern, 'random:') then
```
- The entire implementation of the `random:` filtering tag includes several unnecessary local variables:
  - `TotalGames` could be replaced with `#tResultR`.
  - The return value of `math.random` does not need to be stored in a variable such as `RandomGame`. It can be used immediately `table.insert(tResult, tResultR[math.random(1, #tResultR)])`. Doing it this way also means that there is no need to iterate through `tResultR`.
- There should be an empty argument variant that returns a random game from the current set of already filtered games (i.e. from `atTable`).

BanCrash replied:
These issues are solved right now I think.

Kapiainen commented on Lauhdutin/@Resources/Frontend/GUI.lua:
```diff
@@ -337,6 +387,79 @@ end
+			local RandomGame = 0
+			local TotalGames = 0	
+			for i, game in ipairs(T_ALL_GAMES) do
+				if game["hourstotal"] > 0 then
```
Same issue here as was mentioned in another comment regarding the use of literals as keys instead of using one of the fields of `GAME_KEYS`.

BanCrash replied:
This issue is solved too.

BanCrash commented on Lauhdutin/@Resources/Frontend/GUI.lua:
```diff
@@ -264,8 +264,58 @@ end
 			return
 		end
 		local tResult = {}
+		local HiddenFilter = SKIN:GetVariable('HiddenFilters')
```
I unmade all changes of that feature until better implementation.

BanCrash commented:
I think I solve all the issues, except HiddenFilters that I delete completely. 

Sorry to bother you, it is obvious that I haven't enough knowledge to help you.

BanCrash commented:
Now I think I implemented correctly the hidden filter part. Only remains to add it to every filter already made.

Kapiainen commented:
I'll accept the pull request once you've finalized the implementation of the setting in the filter tags.

BanCrash commented:
Perfect. I will let you know when I'm finished.

BanCrash commented:
Finally I end all the changes. I implemented hidden games feature on all filters, and added all previous Steam filters to Battle.net and Galaxy. I think filtering part is complete now.

Kapiainen commented:
Ok, thanks! I'll have a look through it all, fix the merge conflicts, and accept the pull request when I have the time to do so.

BanCrash commented:
Perfect! Thanks you.
